docs/Cursors.md

```diff
@@ -0,0 +1,287 @@
+- `idx() -> ExprListCursor`: Returns a cursor to the index expression list.
+- `rhs() -> ExprCursor`: Returns a cursor to the right-hand side expression.
+
+
+#### `AssignConfigCursor`
+
+Represents a cursor pointing to a configuration assignment statement of the form:
+```
+config.field = rhs
+```
+
+Methods:
+- `config() -> Config`: Returns the configuration object.
+- `field() -> str`: Returns the name of the configuration field being assigned to.
+- `rhs() -> ExprCursor`: Returns a cursor to the right-hand side expression.
+
+#### `PassCursor`
+
+Represents a cursor pointing to a no-op statement:
+```
+pass
+```
+
+#### `IfCursor`
+
+Represents a cursor pointing to an if statement of the form:
+```
+if condition:
+    body
+```
+or
+```
+if condition:
+    body
+else:
+    orelse
+```
+Returns an invalid cursor if `orelse` isn't present.
+
+Methods:
+- `cond() -> ExprCursor`: Returns a cursor to the if condition expression.
+- `body() -> BlockCursor`: Returns a cursor to the if body block.
+- `orelse() -> Cursor`: Returns a cursor to the else block (if present).
+
+#### `ForCursor`
+
+Represents a cursor pointing to a loop statement of the form:
+```
+for name in seq(0, hi):
+    body
+```
+
+Methods:
+- `name() -> str`: Returns the loop variable name.
+- `lo() -> ExprCursor`: Returns a cursor to the lower bound expression (defaults to 0).
+- `hi() -> ExprCursor`: Returns a cursor to the upper bound expression.
+- `body() -> BlockCursor`: Returns a cursor to the loop body block.
+
+
+#### `AllocCursor`
+
+Represents a cursor pointing to a buffer allocation statement of the form:
+```
+name : type @ mem
+```
+
+Methods:
+- `name() -> str`: Returns the name of the allocated buffer.
+- `mem() -> Memory`: Returns the memory location of the buffer.
+- `is_tensor() -> bool`: Checks if the allocated buffer is a tensor.
+- `shape() -> ExprListCursor`: Returns a cursor to the shape expression list.
+- `type() -> API.ExoType`: Returns the type of the allocated buffer.
+
+
+#### `CallCursor`
+
+Represents a cursor pointing to a sub-procedure call statement of the form:
+```
+subproc(args)
+```
+
+Methods:
+- `subproc()`: Returns the called sub-procedure.
+- `args() -> ExprListCursor`: Returns a cursor to the argument expression list.
+
+
+#### `WindowStmtCursor`
+
+Represents a cursor pointing to a window declaration statement of the form:
+```
+name = winexpr
+```
+
+Methods:
+- `name() -> str`: Returns the name of the window.
+- `winexpr() -> ExprCursor`: Returns a cursor to the window expression.
+
+
+## Cursor Forwarding
+
+When a procedure `p` is transformed into a new procedure `p'` by applying scheduling primitives, any cursors pointing into `p` need to be updated to point to the corresponding locations in `p'`. This process is called *cursor forwarding*.
+
+To forward a cursor `c` from `p` to `p'`, you can use the `forward` method on the new procedure:
+```python
+c' = p'.forward(c)
```

Comment:
I'm not sure if the c-prime, p-prime notation is required by convention, but it causes weird syntax highlighting for me (the assignment is highlighted in the string literal color).